### PR TITLE
[WIP] lsp: add incremental text synchronization via byte ranges

### LINKED ISSUE #13049: [LSP] 'textDocument/didChange' is always sending full buffer, content-ranges would be more efficient
Hello Neovim LSP team,

Spawning this request from a discussion had over at [completion.nvim](https://github.com/nvim-lua/completion-nvim), specifically [this post of mine](https://github.com/nvim-lua/completion-nvim/issues/203#issuecomment-703356444).

As I have noted a few times, I am an [LSC](https://github.com/natebosch/vim-lsc) user, but I am also dipping my toes into Neovim LSP augmented with completion.nvim.

The issue here is that Neovim LSP is always sending the full buffer contents for each `textDocument/didChange` event. This events get issued quite often, for example in insert mode it can sometimes happen after every keypress.

An example of Neovim's LSP sending a `didChange`:

```json
{
  "method": "textDocument/didChange",
  "jsonrpc": "2.0",
  "params": {
    "contentChanges": [
      {
        "text": "class FooBar\n  def initialize\n    @abc = \"abc\"\n    @hello = \"hello\"\n    @help = \"help\"\n  end\n\n  def baz\n    @\n  end\nend\n\nfoo = Foobar.new\n"
      }
    ],
    "textDocument": { "uri": "file:///tmp/foo/foobar.rb", "version": 6 }
  }
}
```

By contrast here is LSC which uses content ranges to describe changes as they occur:

```json
{
  "method": "textDocument/didChange",
  "jsonrpc": "2.0",
  "params": {
    "contentChanges": [
      {
        "range": {
          "end": { "character": 2, "line": 8 },
          "start": { "character": 2, "line": 8 }
        },
        "text": "  @\n  ",
        "rangeLength": 0
      }
    ],
    "textDocument": { "uri": "file:///tmp/foo/foobar.rb", "version": 2 }
  }
}
```

For small text files there likely won't be any issue between full content or content ranges.

But when dealing with multi-thousand line buffers then the current Neovim approach seems inefficient. How will the language server ascertain the actual changes? It must do a de-difference? I do notice that Neovim LSP + completion.nvim can sometimes result in higher CPU usage than LSC when doing the same task.

[This is the Neovim code in question](https://github.com/neovim/neovim/blob/master/runtime/lua/vim/lsp.lua#L813). The comment attached is interesting:

```lua
--[=[ TODO(ashkan) there seem to be problems with the byte_sizes sent by
      -- neovim right now so only send the full content for now. In general, we
      -- can assume that servers *will* support both versions anyway, as there
      -- is no way to specify the sync capability by the client.
      -- See https://github.com/palantir/python-language-server/commit/cfd6675bc10d5e8dbc50fc50f90e4a37b7178821#diff-f68667852a14e9f761f6ebf07ba02fc8 for an example of pyls handling both.
      --]=]
```

Full content `didChange` *could* be a performance bottleneck, it would likely be more efficient to use content ranges.

Best regards.

### PR DESCRIPTION
**Pull request by mjlbach**
Closes #13049

Continuing @tjdevries work. This now works but some tests are failing.

### CODE REVIEW

File: runtime/lua/vim/lsp.lua [hunk 1]
```diff
@@ -811,78 +811,119 @@ local function once(fn)
...
       return {
         range = {
-          start = { line = firstline, character = 0 };
-          ["end"] = { line = lastline, character = 0 };
+          -- TODO: UTF??
+          -- TODO: Is it old end or new end?
+          start = { line = start_row, character = start_column - 1 };
+          ["end"] = { line = start_row+old_end_row, character = start_column+old_end_column - 1};
         };
-        rangeLength = length;
```
Comment (bfredl) [inline runtime/lua/vim/lsp.lua:852]: isn't `rangeLength` still needed?

Comment (mjlbach) [reply runtime/lua/vim/lsp.lua:852]: That was recently deprecated https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_didChange

Comment (mjlbach) [reply runtime/lua/vim/lsp.lua:852]:
```
	/**
	 * The optional length of the range that got replaced.
	 *
	 * @deprecated use range instead.
	 */
	rangeLength?: uinteger;
```

Comment (bfredl) [reply runtime/lua/vim/lsp.lua:852]: And servers (the common ones) have been updated to not require it?

Comment (mjlbach) [reply runtime/lua/vim/lsp.lua:852]: I don't think sumneko, RLS, HLS, or vscode-language server node (pyright, tsserver, yamlls, etc) require it based on a cursory look at the sources, but I can add it for backwards compatibility

Comment (bfredl) [reply runtime/lua/vim/lsp.lua:852]: Well in that case we can wait and see if someone complains :)

**Comment by mjlbach** [comment]
@bfredl probably out of scope of this PR, but it would be nice to add nvim_get_bytes to the API, which would fetch the buffer text based on the return result of on_bytes

**Comment by tjdevries** [comment]
hehe yes, the PR is trivisal with `nvim_buf_get_bytes()` :)

**Comment by tjdevries** [comment]
( @bfredl  ^^ )

**Comment by mjlbach** [comment]
I'm also not positive on how the range start/end are handled for deleted/added multiline, which is the only other holdup (and mildly related)

**Comment by mjlbach** [comment]
The more I dig into this, the more I think we have to add `nvim_buf_get_bytes()` I started looking into this, but some of the old vim functions operating of buf offsets are kind of opaque to me.

image

I keep running into issues, as wihtout a way to reliably yield the text from the buf offset, it's very unclear what to do here. I thought I could get by with just the column/row offsets, but that's not the case.

**Comment by bfredl** [comment]
hmm byte offsets should be irrelevant for LSP if `rangeLenght` is not used. `nvim_buf_get_bytes` would be _convenient_ but not change anything fundamentally (until we get ultralong line support)

**Comment by mjlbach** [comment]
It's more that I was planning on re-using some of the functions that are used to calculate byte_offset to get the text associated with a byte range. As it stands currently, I don't believe that having the column/row changes are going to be enough to get the correct byte range to send to the server, and we don't have a good way currently to use byte_offset and new_byte_end_length from the on_bytes callback to get the range.

**Comment by bfredl** [comment]
I don't see how byte ranges are relevant at all. start col/row and old_end col/row defines the range that is replaced and sent as a `range` object to the server. while start col/row and new_end col/row defines the text that should be extracted from the buffer and sent to the server. To be fully UTF-16 compliant effectively, we would need to access the replaced lines in the undo header (undolevel>=1 is an acceptable requirement for enabling incremental sync) to convert column positions. But regardless, the global byte offsets never enter into the picture, as far as I can see.

**Comment by bluz71** [comment]
> I don't see how byte ranges are relevant at all. start col/row and old_end col/row defines the range that is replaced and sent as a `range` object to the server.

If I understand your post correctly (and I may not), you appear to query the need for this PR at all since line range changes already exists in Neovim and should that be enough. If so, let me note the why of this PR.

In an ideal world yes, line ranges should be enough.

But issue #13049 notes that at a minimum the Ruby language server (Solargraph) will not complete after a trigger character if full bufffer or line based content changes are sent. Completion basically does not work.

If on the other-hand minumal character-ranges are sent (as proposed in this PR) then completion will succeed.

One may argue that the Ruby language server should not do this. But one may also counter by stating that if the Ruby language server does this then it may be the case that certain other language servers may also behave this way. I have not tested all the language servers that exist, but I know for certain that one of the important ones behaves as such.

I also know for certain that the [vim-lsc](https://github.com/natebosch/vim-lsc) and [vim-lsp](https://github.com/prabirshrestha/vim-lsp) plugins send character range based content changes (not line changes). And I am almost 100% certain that Visual Studio Code also sends minimal character ranges (since Ruby completion works and that the Ruby language server is written specifically for VSCode).

I believe the intention of the LSP content ranges is to encapsulate a minimal edit in the smallest possible request. It seems to be the case that many LSP clients do exactly that.

Apologies if I have misinterpreted your post @bfredl.

Best regards.

**Comment by mjlbach** [comment]
@bluz71 Bfredl is just telling me we don't need byte ranges implemented in nvim_api (on the c side) for this PR, which I think is true, but I also think would lead to a cleaner implemenation.I think everyone is in agreement that we should send the range based changes to the language server.

**Comment by bluz71** [comment]
Thanks @mjlbach, I appreciate the explanation. Hopefully no offence taken.

Side note, I will now be testing your latest branch and report back over the next hour or so. Stay tuned.

**Comment by mjlbach** [comment]
No worries! Just wanted to clear things up. Sorry for all the back and forth on this, the spec doesn't make it clear what it expects for ranges (or at least it didn't when I read it!) I should probably just set up vim-lsc myself to a/b test , but it's been forever since I used another language client 😆

**Comment by mjlbach** [comment]
@tjdevries I'm not going to lie, I basically randomly tried different range combinations until the language server results looked correct... I'm not sure if there's a better way to test this (maybe there's a language server that will let us print out a representation of the document after each edit?)

Anyways, I think it at least runs now! I'll go ahead and add a test unless you see something obviously wrong 😆

**Comment by mjlbach** [comment]
Also not quite sure what is causing the failing tests, seems like they are all in lsp_spec.lua, but we shouldn't have touched anything with the full_range changes...

**Comment by bfredl** [comment]
@bluz71 the values I referred to (row/col pairs) only exists in `on_bytes` to start with. The only benefit of `on_lines` is automatic codeunit/codepoint lenght calculation, and that is irrelevant without `rangeLength` .

@mjlbach even simpler. I just noted the six row/col values should be enough to calculate the update sent to LSP. Neither neovim nor the LSP protocol seems to fundamentally care about global byte offsets/lengths (but TS does, which is why added them), so using them here seems only a distraction to me IIUC.

**Comment by mjlbach** [comment]
I just thought that nvim_buf_get_byte(bufnr, byte_offset - new_end_byte, byte_offset) (not sure if these are the correct ranges) would be simpler/read better than what I'm current doing in:  

https://github.com/neovim/neovim/pull/13979/files#diff-c6a4160a416b22cd1fd833b8f852c4fbaf1f25adf75053e82a1025aa2c607ac9R814-R834

File: runtime/lua/vim/lsp.lua [hunk 2]
```diff
@@ -811,78 +811,117 @@ local function once(fn)
...
-       table.insert(lines, '')
+    local incremental_changes = once(function(_client)
+      local text = buf_range_to_text(bufnr, {
+                      start_row = start_row,
+                      start_column = start_column,
+                      new_end_row=new_end_row,
+                      new_end_column = new_end_column,
+                      })
+
+      if new_end_byte_length < old_end_byte_length then
```
Comment (bfredl) [inline runtime/lua/vim/lsp.lua:880]: This is not correct. `text` would be empty only when `new_end_byte_length` is zero, but if `buf_range_to_text` is correct that will already be the case.

Comment (mjlbach) [reply runtime/lua/vim/lsp.lua]: Fixed, thank you

File: runtime/lua/vim/lsp.lua [hunk 3]
```diff
@@ -811,78 +811,113 @@ local function once(fn)
...
 
+--@private
+--- converts retrieves buffer text based on row and column range
+---
+--@param bufnr integer Buffer handle
+--@param range table table containing start_row, start_column, new_end_row, and new_end_column from on_bytes callback
+--@returns string text from buffer
+local function buf_range_to_text(bufnr, range)
+  local lines = vim.api.nvim_buf_get_lines(bufnr, range.start_row, range.start_row + range.new_end_row + 1, false)
+  for idx = 1, #lines do
```
Comment (bfredl) [inline runtime/lua/vim/lsp.lua:822]: I don't think `\n` on final newline should be there. We are extracting one extra line to get the `new_end_column` part of the final line. so we can use `table.concat(lines, '\n')` in the end instead.

Comment (mjlbach) [reply runtime/lua/vim/lsp.lua:822]: I had this originally, but it wasn't representing the test correctly for single empty line additions (like `o`)

Comment (bfredl) [reply runtime/lua/vim/lsp.lua:822]: Hmm are you sure? with `o` new_end_row should be 1 so we get one more line, is this not the case?

Comment (mjlbach) [reply runtime/lua/vim/lsp.lua:822]: Here's what happens when I use table.concat

https://user-images.githubusercontent.com/13316262/108682719-84b7dc80-74a5-11eb-84f5-c8263bd297ba.mov

Comment (mjlbach) [reply runtime/lua/vim/lsp.lua:822]: This is what happens with the PR as it currently is:

https://user-images.githubusercontent.com/13316262/108683109-060f6f00-74a6-11eb-8366-114ddd70625b.mov

File: runtime/lua/vim/lsp.lua (hunk 2 above)
Comment (mjlbach) [reply runtime/lua/vim/lsp.lua]: Wait, I think we actually need this. I'm getting errors now when doing live editing.

Comment (bfredl) [reply runtime/lua/vim/lsp.lua]: Well you can check if `new_end_row` and `new_end_column` is both zero in `buf_range_to_text` and then return empty string even before API-call.

File: runtime/lua/vim/lsp.lua (hunk 3 above)
Comment (bfredl) [reply runtime/lua/vim/lsp.lua:822]: That's weird because the updates for the latter looks wrong: it inserts two extra lines upon undo.

Comment (mjlbach) [reply runtime/lua/vim/lsp.lua:825]: It's definitely still wrong, I'm just trying to figure out exactly how it is wrong 😆 Adding and undoing 3 lines breaks it

**Comment by mjlbach** [comment]
Hmm still getting a ton of errors that popup with undo or paste over multiple lines, or paste that makes a line shorter combined with undo

**Comment by mjlbach** [comment]
@bfredl Here is the minimum text case:

1. Make a file
```
local function main ()
  return "abc"
end
```

2. Store the following macros:
```
:let @a="jllllllhvllllllyvlllllpu"
:let @b="jlllllvllllllyvlllllpx"
```

3. Execute @b, you'll see 
```
{
  byte_offset = 34,
  lines = { '  return "abc"' },
  new_end_byte_length = 0,
  new_end_column = 0,
  new_end_row = 0,
  old_end_byte_length = 1,
  old_end_column = 1,
  old_end_row = 0,
  range = {
    end = {
      character = 12,
      line = 1
    },
    start = {
      character = 11,
      line = 1
    }
  },
  start_column = 11,
  start_row = 1,
  text = ""
}
{
  byte_offset = 28,
  lines = { '  return "abbc"' },
  new_end_byte_length = 7,
  new_end_column = 7,
  new_end_row = 0,
  old_end_byte_length = 0,
  old_end_column = 0,
  old_end_row = 0,
  range = {
    end = {
      character = 5,
      line = 1
    },
    start = {
      character = 5,
      line = 1
    }
  },
  start_column = 5,
  start_row = 1,
  text = 'urn "ab'
}
{
  byte_offset = 28,
  lines = { '  retbc"' },
  new_end_byte_length = 0,
  new_end_column = 0,
  new_end_row = 0,
  old_end_byte_length = 6,
  old_end_column = 6,
  old_end_row = 0,
  range = {
    end = {
      character = 11,
      line = 1
    },
    start = {
      character = 5,
      line = 1
    }
  },
  start_column = 5,
  start_row = 1,
  text = ""
}
```

4. now execute @a 
```
{
  byte_offset = 28,
  lines = { '  return "abc"' },
  new_end_byte_length = 6,
  new_end_column = 6,
  new_end_row = 0,
  old_end_byte_length = 7,
  old_end_column = 7,
  old_end_row = 0,
  range = {
    end = {
      character = 6,
      line = 1
    },
    start = {
      character = 5,
      line = 1
    }
  },
  start_column = 5,
  start_row = 1,
  text = ""
}
{
  byte_offset = 28,
  lines = { '  return "abbc"' },
  new_end_byte_length = 7,
  new_end_column = 7,
  new_end_row = 0,
  old_end_byte_length = 0,
  old_end_column = 0,
  old_end_row = 0,
  range = {
    end = {
      character = 5,
      line = 1
    },
    start = {
      character = 5,
      line = 1
    }
  },
  start_column = 5,
  start_row = 1,
  text = 'urn "ab'
}
{
  byte_offset = 28,
  lines = { '  retbc"' },
  new_end_byte_length = 0,
  new_end_column = 0,
  new_end_row = 0,
  old_end_byte_length = 6,
  old_end_column = 6,
  old_end_row = 0,
  range = {
    end = {
      character = 11,
      line = 1
    },
    start = {
      character = 5,
      line = 1
    }
  },
  start_column = 5,
  start_row = 1,
  text = ""
}

```

Note these are slightly different in terms of cursor positioning, but the representation of the buffer on the server side is corrupted in the latter, but not in the former. This is with the latest version of this PR.

**Comment by mjlbach** [comment]
@bluz71 Do you want try editing with this for a bit? No need for the vim lsc test

**Comment by mjlbach** [comment]
ah nvm, found a bug.

**Comment by mjlbach** [comment]
@tjdevries I have to put this on hold (possibly indefinitely), I think I'm going a bit crazy trying to fix the corner cases 😕

**Comment by bluz71** [comment]
> @bluz71 Do you want try editing with this for a bit? No need for the vim lsc test

Do you still want me to test?

>  I have to put this on hold (possibly indefinitely)

It sounds like with your latest post that this PR will be shelved. If so, it may not be worth my time to test.

Let me know how you want to proceed.

P.S. I appreciate your work on this item. Hopefully we can get there one day since I am sure it will improve Neovim's LSP.

**Comment by mjlbach** [comment]
The PR won't be shelved, and I agree it's useful, I just think I hit the 80/20 rule hard. It's mostly there, and others are free to pick it up before I swing back to it.

**Comment by bluz71** [comment]
Ok I understand, hopefully you and someone picks this up again.

When you need me to test don't hesitate to ask.

**Comment by mjlbach** [comment]
I lied and kept working on this. I wrote something to dump out the internal state of the julia language server, which sadly doesn't apparently do incremental sync (whoops), but I noticed one of the main issues I ran into (moving lines) actually reproduces the bug on full text change (weird)

https://user-images.githubusercontent.com/13316262/108959856-df7a4100-7629-11eb-98d0-e1d1f6789ed7.mov


Note the duplication on server side of the line after each move action

**Comment by mjlbach** [comment]
This is useful for debugging purposes:

```
diff --git a/crates/rust-analyzer/src/main_loop.rs b/crates/rust-analyzer/src/main_loop.rs
index 2829d5970..822d9b5fb 100644
--- a/crates/rust-analyzer/src/main_loop.rs
+++ b/crates/rust-analyzer/src/main_loop.rs
@@ -13,6 +13,7 @@ use ide_db::base_db::VfsPath;
 use lsp_server::{Connection, Notification, Request, Response};
 use lsp_types::notification::Notification as _;
 use vfs::ChangeKind;
+use std::fs;
 
 use crate::{
     config::Config,
@@ -594,7 +595,7 @@ impl GlobalState {
                     let file_id = vfs.file_id(&path).unwrap();
                     let mut text = String::from_utf8(vfs.file_contents(file_id).to_vec()).unwrap();
                     apply_document_changes(&mut text, params.content_changes);
-
+                    fs::write("/Users/michael/rust_analyzer_output.rs", text.clone())?;
                     // The version passed in DidChangeTextDocument is the version after all edits are applied
                     // so we should apply it before the vfs is notified.
                     doc.version = params.text_document.version;

```

**Comment by bfredl** [comment]
> @bfredl Here is the minimum test case:
> 
> 
> Note these are slightly different in terms of cursor positioning, but the representation of the buffer on the server side is corrupted in the latter, but not in the former. This is with the latest version of this PR.

Hmm the internal byte updates looks correct (tested with nsync.nvim). the latest event in `@b` is just a single byte delete, so I am not sure why it would fail.

**Comment by mjlbach** [comment]
This PR is on hold pending updates to on_bytes to support more incremental buffer updates necessary (on_bytes is currently heavily optimized for tree-sitter, but not granular enough for LSP)

**Pull request closed without merging**